### PR TITLE
[aggregator] Fix metric annotations being dropped on conversion to/from proto

### PR DESCRIPTION
**Pull request by vpranckaitis**
**What this PR does / why we need it**:

Fixes metric annotations being dropped on the wire transfer to aggregator, which resulted in aggregated metrics having no annotations, too. Also, updated aggregator integration tests to use non-`nil` annotations.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/aggregator/aggregation/gauge.go [hunk 1]
```diff
@@ -64,6 +64,17 @@ func (g *Gauge) Update(timestamp time.Time, value float64, annotation []byte) {
...
 
+	// Keep the last annotation which was set.
+	if len(annotation) > 0 {
+		if cap(g.annotation) < len(annotation) {
+			// Twice as long in case another one comes in
+			// and we could avoid realloc as long as less than this first alloc.
+			g.annotation = make([]byte, 0, 2*len(annotation))
+		}
+		// Reuse any previous allocation while taking a copy.
+		g.annotation = append(g.annotation[:0], annotation...)
```
Comment (linasm) [inline src/aggregator/aggregation/gauge.go:75]: Can we be sure reusing of slice is safe here? I.e. isn't there a case when this reusing will invalidate the slice which has been passed to a protobuf struct (by reference), but not marshalled yet?

Comment (vpranckaitis) [reply src/aggregator/aggregation/gauge.go:75]: While I don't have the whole picture and it might be useful to loop in someone else, but after a quick look-around it seems that writes and reads/flushes are mutually exclusive. E.g. `X ∈ {Counter, Gauge, Timer}` is being wrapped in `lockedXAggregation`:

https://github.com/m3db/m3/blob/584f5d4bd03f82153a932a5e9288546c22343598/src/aggregator/aggregator/gauge_elem_gen.go#L43-L49

which is locked during updates

https://github.com/m3db/m3/blob/584f5d4bd03f82153a932a5e9288546c22343598/src/aggregator/aggregator/gauge_elem_gen.go#L151-L157

and during flushes

https://github.com/m3db/m3/blob/584f5d4bd03f82153a932a5e9288546c22343598/src/aggregator/aggregator/gauge_elem_gen.go#L251-L273

Also, the things inside seem to be handled in synchronous way and the `Counter` / `Gauge` / `Timer` doesn't seem to be reused afterwards.

**Review by linasm — APPROVED** [review]

File: src/metrics/generated/proto/metricpb/metric.proto [hunk 2]
```diff
@@ -34,16 +34,19 @@ enum MetricType {
...
   bytes id = 1;
   repeated double values = 2;
+  bytes annotation = 3;
 }
 
 message Gauge {
   bytes id = 1;
   double value = 2;
+  bytes annotation = 3;
 }
```
Comment (vdarulis) [inline src/metrics/generated/proto/metricpb/metric.proto:51]: if you are changing protos, make sure to update `src/metrics/encoding/protobuf/reset.go` to reuse the new fields, too.

File: src/aggregator/aggregation/timer.go [hunk 3]
```diff
@@ -81,6 +71,17 @@ func (t *Timer) AddBatch(timestamp time.Time, values []float64) {
...
+	// Keep the last annotation which was set.
+	if len(annotation) > 0 {
+		if cap(t.annotation) < len(annotation) {
+			// Twice as long in case another one comes in
+			// and we could avoid realloc as long as less than this first alloc.
+			t.annotation = make([]byte, 0, 2*len(annotation))
+		}
+		// Reuse any previous allocation while taking a copy.
+		t.annotation = append(t.annotation[:0], annotation...)
+	}
```
Comment (vdarulis) [inline src/aggregator/aggregation/timer.go:84]: just make a helper method, this seems to be sprinkled all over the codebase.
are you sure that the reallocation sizing is even worth it? might make memory usage higher than really needed.

**Review by robskillington — APPROVED** [review]
LGTM
